### PR TITLE
rework how we handle outlives relationships

### LINKED ISSUE #53121: [nll] spurious outlives requirement in shred-0.7.0
The following snippet, extracted from the crate shred v0.7.0, fails to compile with NLL, **but only when compiled with `--crate-type lib`**: 

```rust
// NB You must compile with `--crate-type lib` to see the failure!

#![feature(nll)]

pub trait Accessor: Sized {
}

pub enum AccessorCow<'a, 'b, T>
where
    AccessorTy<'a, T>: 'b,
    T: System<'a> + ?Sized,
'a: 'b,
{
    /// A reference to an accessor.
    Ref(&'b AccessorTy<'a, T>),
    /// An owned accessor.
    Owned(AccessorTy<'a, T>),
}

type AccessorTy<'a, T> = <<T as System<'a>>::SystemData as DynamicSystemData<'a>>::Accessor;

pub trait System<'a> {
    type SystemData: DynamicSystemData<'a>;
}

pub trait SystemData<'a> {
}

pub trait DynamicSystemData<'a> {
    /// The accessor of the `SystemData`, which specifies the read and write dependencies and does
    /// the fetching.
    type Accessor: Accessor;
}

fn main() { }
```

yields

```
> rustc +nightly ~/tmp/reduce.rs -Zborrowck=mir -Ztwo-phase-borrows --crate-type lib

error[E0309]: the parameter type `T` may not live long enough
  --> /home/nmatsakis/tmp/reduce.rs:13:5
   |
13 |     Ref(&'b AccessorTy<'a, T>),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `T: 'b`...

error[E0309]: the parameter type `T` may not live long enough
  --> /home/nmatsakis/tmp/reduce.rs:15:5
   |
15 |     Owned(AccessorTy<'a, T>),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `T: 'b`...
```

I'm not entirely sure what is going on here, but it seems very fishy, particularly since it is specific to `--crate-type lib`.

### LINKED ISSUE #53789: [NLL] Possible proptest-arbitrary-0.2.2 regression
Filing for investigation.

The recent crater run found an interesting case in `proptest-arbitrary` 0.2.2, which used to compile with AST borrowck.

[Crater logs.](https://crater-reports.s3.amazonaws.com/pr-53426/a5d98b025f62ace80fe9d2a15781f900fc42515e-alt/reg/proptest-arbitrary-0.2.2/log.txt)

There are 2 reproductions: 
- the first collates the necessary parts from `proptest-arbitrary` and `proptest` and reproduces faithfully the messages from the crater run. [Playground](https://play.rust-lang.org/?gist=c990af931ab56d93207efcade0ffd01b&version=nightly&mode=debug&edition=2015).
- the second is a bit more minimal, while still being accepted by AST borrowck. It is however, in an effort to simplify the code, partially automatically reduced to focus on only one of these two identical errors. [Playground](https://play.rust-lang.org/?gist=1f1d03d4f8280b8843594cba668c9db4&version=nightly&mode=debug&edition=2015).

### PR DESCRIPTION
**Pull request by nikomatsakis**
When we encounter an outlives relationship involving a projection, we use to over-constrain in some cases with region constraints. We also used to evaluate whether the where-clauses in the environment might apply **before** running inference.

We now avoid doing both of those things:

- If there are where-clauses in the environment that might be useful, we add no constraints.
- After inference is done, we check if we wound up inferring values compatible with the where-clause, and make use of them if so.

I realize now that this PR includes some meandering commits and refactorings from when I expected to go in a different direction. If desired, I could try to remove some of those.

Fixes #53121 
Fixes #53789 

r? @pnkfelix

### CODE REVIEW

**Comment by pnkfelix** [comment]
Aww tidy failure ...

File: src/librustc/infer/mod.rs [hunk 1]
```diff
@@ -80,6 +80,16 @@ pub type Bound<T> = Option<T>;
 pub type UnitResult<'tcx> = RelateResult<'tcx, ()>; // "unify result"
 pub type FixupResult<T> = Result<T, FixupError>; // "fixup result"
 
+/// A flag that is given when running region resolution: if true, it
+/// indicates that we should not report the region errors to the user
+/// if NLL is enabled, since NLL will also detect them (and do a
+/// better job of it).
```
Comment (pnkfelix) [inline src/librustc/infer/mod.rs]: 1. Should this comment include the special role of `borrowck=migrate`?

2. Instead of checking for the borrowck=mode after seeing `UnlessNll(true)`, *maybe* we should revise how `UnlessNll` itself is initialized based on inspecting that `BorrowckMode`?
  * Its up to you; I just know that I initially found the name `UnlessNll` confusing, given that we *will* emit the errors under the migration mode... But of course that was inherited from the previous code...

File: src/librustc/infer/outlives/verify.rs [hunk 2]
```diff
@@ -74,10 +74,18 @@ impl<'cx, 'gcx, 'tcx> VerifyBoundCx<'cx, 'gcx, 'tcx> {
...
+    ///
+    /// This is an "approximate" check -- it may not find all
+    /// applicable bounds, and not all the bounds it returns can be
+    /// relied upon. In particular, this check ignores region
+    /// identity.  So, for example, if we have `<T as
+    /// Trait<'0>>::Item` where `'0` is a region variable, and the
+    /// user has `<T as Trait<'a>>::Item: 'b` in the environment, then
+    /// the clause from the environment only applies if `'0 = 'a`,
+    /// which we don't know yet. But we would still include `'b` in
+    /// this list.
```
Comment (pnkfelix) [inline src/librustc/infer/outlives/verify.rs]: I'm going through the PR commit-by-commit, so this question may get answered later on, but I don't want to forget to ask it: What is the intended use of this check, where it sounds like it is incomplete on two axes...?

My current hypothesis is that this check is used to provide diagnostics in scenarios where we are comfortable giving feedback suggesting the user *might* need to satisfy some constraint, when the reality is that they might be able to sidestep the constraint... That probably didn't make sense. let me try again

Here's a different version of my question: You have given a concrete example where the list returned by this function would include `'b`, even though that bound may not apply, since it depends on the whether it is true that `'0 = 'a`. So: Do you have an example of a diagnostic where as a user one can *see* the effect of `'b` being on this list?

Comment (pnkfelix) [reply src/librustc/infer/outlives/verify.rs]: My question might be answered by cebf83502fd7c57dd2a7d1481b00d93ec03bab0e

Comment (pnkfelix) [reply src/librustc/infer/outlives/verify.rs]: (Though commit cebf83502fd7c57dd2a7d1481b00d93ec03bab0e only has tests being updated that are observing internal region representations... I *think* some of tests still show side-effects that end-users can observe without using `-Z` flags, but I *hope* that by the time I get to the end of the commit series, there will be a concrete test that demonstrates the effect without `-Z` flags.)

File: src/test/ui/nll/ty-outlives/projection-where-clause-none.rs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#![feature(nll)]
+
+// Test that we are NOT able to establish that `<T as
```
Comment (pnkfelix) [inline src/test/ui/nll/ty-outlives/projection-where-clause-none.rs]: why is there no `projection-where-clause-none.stderr` in this PR?

Comment (pnkfelix) [reply src/test/ui/nll/ty-outlives/projection-where-clause-none.rs]: (it seems like a bunch of `.stderr` files have been left out of this PR?)

**Comment by pnkfelix** [comment]
Okay this all seems fine.

r=me once you've placated travis, which presumably will require tidying and adding some missing `.stderr` files.

File: src/librustc/infer/mod.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc/infer/mod.rs]: I feel like if we do the checking *before* initializing `UnlessNll`, I would just call it `Ignore` or something like that, since at that point it is encoding whether to ignore the error and nothing else, right?

(Right now it's saying "is this the sort of error we should ignore if NLL is enabled" — maybe I can improve the comment this way?)

Comment (nikomatsakis) [reply src/librustc/infer/mod.rs]: I'm not inclined to move the logic earlier because it's created many places -- well, I guess I can just have a constructor like `RegionErrors::suppress_if_nll()` or something?

Comment (nikomatsakis) [reply src/librustc/infer/mod.rs]: ok, did that, and it seems nice. will push a commit.

**Comment by nikomatsakis** [comment]
@pnkfelix 

> What is the intended use of this check, where it sounds like it is incomplete on two axes...?

I'm pulling the answer to this out because I will rebase and your comments will likely be lost. The idea of the "approximate" check is that we find out: "Are there any where clauses that might **potentially** be useful once region inference is done, presuming we inferred the various region values to the right values?"

It's not really meant to be conservative in "two dimensions". That is, `projection_approx_declared_bounds_from_env` ought to return anything that might **potentially** be useful. I suspect this is not presently true, though, around bound regions and a few other edge cases (e.g., I think we by and large ignore things like `for<'a> T::Proj<'a>: 'a`, which is really a bug, although a pre-existing one; we also are too conservative I think around equality of types like `for<'a> fn(&'a u32)` vs `for<'b> fn(&'b u32)`. Also a pre-existing bug.)

I might be able to fix those cases where we are "overly" conservative though via a more thorough erasure. I think erasing regions today preserves the lifetime bounds, but we would ideally look for an "equality" check that just **completely** ignores region equality (even around binders). I was working towards that but never finished it.

File: src/test/ui/nll/ty-outlives/projection-where-clause-none.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/test/ui/nll/ty-outlives/projection-where-clause-none.rs]: hmm, curious, I have no such files locally...? I'll fix, anyway

**Comment by nikomatsakis** [comment]
filed https://github.com/rust-lang/rust/issues/54572 as a follow-up issue
